v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/bigtablechangestreamstopubsub/schemautils/PubSubUtils.java

```diff
@@ -0,0 +1,388 @@
+    changelogEntryMessage.setSourceTable(
+        (String) FORMATTERS.get(PubSubFields.SOURCE_TABLE).format(this, changeJsonParsed));
+
+    Publisher publisher = null;
+    ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
+
+    Encoder encoder = null;
+    switch (messageEncoding) {
+      case "BINARY":
+        encoder = EncoderFactory.get().directBinaryEncoder(byteStream, /* reuse= */ null);
+        break;
+      case "JSON":
+        encoder =
+            EncoderFactory.get().jsonEncoder(ChangelogEntryMessage.getClassSchema(), byteStream);
+        break;
+      default:
+        final String errorMessage =
+            "Invalid message encoding: "
+                + messageEncoding
+                + ". Supported output formats: JSON, BINARY";
+        throw new IllegalArgumentException(errorMessage);
+    }
+    changelogEntryMessage.customEncode(encoder);
+    encoder.flush();
+
+    // Publish the encoded object as a Pub/Sub message.
+    ByteString data = ByteString.copyFrom(byteStream.toByteArray());
+    PubsubMessage message = PubsubMessage.newBuilder().setData(data).build();
+    return message;
+  }
+
+  public PubsubMessage mapChangeJsonStringToPubSubMessageAsJson(String changeJsonSting)
+      throws Exception {
+    String messageEncoding = this.getDestination().getMessageEncoding();
+    ChangelogEntryMessageJson changelogEntryMessageJson = new ChangelogEntryMessageJson();
+    JSONObject changeJsonParsed = new JSONObject(changeJsonSting);
+
+    if (this.destination.getUseBase64Rowkey()) {
+      changelogEntryMessageJson.setRowKey(
+          (String)
+              FORMATTERS.get(PubSubFields.ROW_KEY_STRING_BASE64).format(this, changeJsonParsed));
+    } else {
+      changelogEntryMessageJson.setRowKey(
+          (String) FORMATTERS.get(PubSubFields.ROW_KEY_STRING).format(this, changeJsonParsed));
+    }
+    changelogEntryMessageJson.setModType(
+        ModType.valueOf(
+            (String) FORMATTERS.get(PubSubFields.MOD_TYPE).format(this, changeJsonParsed)));
+    changelogEntryMessageJson.setIsGC(
+        (Boolean) FORMATTERS.get(PubSubFields.IS_GC).format(this, changeJsonParsed));
+    changelogEntryMessageJson.setTieBreaker(
+        (Integer) FORMATTERS.get(PubSubFields.TIEBREAKER).format(this, changeJsonParsed));
+    changelogEntryMessageJson.setColumnFamily(
+        (String) FORMATTERS.get(PubSubFields.COLUMN_FAMILY).format(this, changeJsonParsed));
+    changelogEntryMessageJson.setCommitTimestamp(
+        (Long) FORMATTERS.get(PubSubFields.COMMIT_TIMESTAMP).format(this, changeJsonParsed));
+    if (this.destination.getUseBase64ColumnQualifiers()) {
+      changelogEntryMessageJson.setColumn(
+          (String)
+              FORMATTERS.get(PubSubFields.COLUMN_STRING_BASE64).format(this, changeJsonParsed));
+    } else {
+      changelogEntryMessageJson.setColumn(
+          (String) FORMATTERS.get(PubSubFields.COLUMN_BYTES).format(this, changeJsonParsed));
```

Comment:
This seems to cast byte[] to String.  On line 249 we have the same logic looking correct for ROW_KEY_STRING, but for some reason we use COLUMN_BYTES. Asymmetry!

Reply:
When casting (bytes[]) to String, the value will be something like this:

"byte[@92u912u43"

Reply:
Any object can be casted to String, a method .toString will be implicitly called which exists for all objects in Java, however, not all classes implement this method. Arrays implement it in that weird way which prints class name "byte[" and address